src/common/Timer.cc

```diff
@@ -94,7 +95,16 @@ void SafeTimer::timer_thread()
       
       if (!safe_callbacks)
 	lock.Unlock();
-      callback->complete(0);
+
+      if (cycle_events.count(callback)){
+        callback->finish(0);
+        cycle_events_t::iterator c_iter = cycle_events.find(callback);
```

Comment:
Use cycle_events.find(callback) once to check whether item exists or not (returns cycle_events.end() in that case), then reuse that iterator here.